File: Java-Core/src/main/java/presto/expression/ModuloExpression.java
package presto.expression;

import presto.error.PrestoError;
import presto.error.SyntaxError;
import presto.runtime.Context;
import presto.type.IType;
import presto.utils.CodeWriter;
import presto.value.IValue;

public class ModuloExpression implements IExpression {

	IExpression left;
	IExpression right;
	
	public ModuloExpression(IExpression left, IExpression right) {
		this.left = left;
		this.right = right;
	}
	
	@Override
	public void toDialect(CodeWriter writer) {
		left.toDialect(writer);
		writer.append(" % ");
		right.toDialect(writer);
	}
	
	@Override
	public IType check(Context context) throws SyntaxError {
		IType lt = left.check(context);
		IType rt = right.check(context);
		return lt.checkModulo(context, rt);
	}
	
	@Override
	public IValue interpret(Context context) throws PrestoError {
		IValue lval = left.interpret(context);
		IValue rval = right.interpret(context);
        return lval.Modulo(context, rval);
	}
}

File: Java-CodeServer/src/main/resources/js/prompto/expression/ModuloExpression.js
var SyntaxError = require("../error/SyntaxError").SyntaxError;
var Value = require("../value/Value").Value;

function ModuloExpression(left, right) {
	this.left = left;
	this.right = right;
	return this;
}

ModuloExpression.prototype.toString = function() {
	return this.left.toString() + " % " + this.right.toString();
};

ModuloExpression.prototype.toDialect = function(writer) {
    this.left.toDialect(writer);
    writer.append(" % ");
    this.right.toDialect(writer);
};

ModuloExpression.prototype.check = function(context) {
	var lt = this.left.check(context);
	var rt = this.right.check(context);
	return lt.checkModulo(context, rt);
};

ModuloExpression.prototype.interpret = function(context) {
	var lval = this.left.interpret(context);
	var rval = this.right.interpret(context);
	return lval.Modulo(context, rval);
};

exports.ModuloExpression = ModuloExpression;
